HouseRules_Essentials/Rules/LevelSequenceOverriddenRule.cs

```diff
@@ -27,37 +19,53 @@ public sealed class LevelSequenceOverriddenRule : Rule, IConfigWritable<List<str
+                prefix: new HarmonyMethod(typeof(LevelSequenceOverriddenRule), nameof(LevelSequenceConfiguration_GetSequenceDefinition_Prefix)));
         }
 
-        protected override void OnDeactivate(GameContext gameContext)
+        /// <remarks>
+        /// Sets a safe sequence definition even if the active game type does not have one that extends to the current level.
+        /// </remarks>
+        private static bool LevelSequenceConfiguration_GetSequenceDefinition_Prefix(ref SequenceDefinition __result, int index, LevelSequence.GameType gameType)
         {
-            ReplaceExistingProperties(_originals);
+            var gameContext = Traverse.Create(typeof(GameHub)).Field<GameContext>("gameContext").Value;
+            var sequenceDefinitions = gameContext.levelSequenceConfiguration.sequenceDefinitions.GetSequenceFromId(gameType, out _);
+            if (index >= 0 && index < sequenceDefinitions.Length)
+            {
+                return true;
+            }
+
+            __result = gameContext.levelManager.GetLevelSequence().CurrentLevelIsLastLevel
+                ? sequenceDefinitions[sequenceDefinitions.Length - 1]
+                : sequenceDefinitions[sequenceDefinitions.Length - 3];
+
+            return false;
         }
 
         /// <summary>
         /// Replaces LevelSequence levels with predefined list.
         /// </summary>
-        /// <returns>List of previous LevelSequnece levels that are now replaced.</returns>
-        private static List<string> ReplaceExistingProperties(List<string> replacements)
+        /// <returns>List of previous LevelSequence levels that are now replaced.</returns>
+        private static List<string> ReplaceExistingProperties(List<string> replacements, GameContext gameContext)
         {
-            var levelManager = Resources.FindObjectsOfTypeAll<LevelManager>().First();
-            var levelSequence = Traverse.Create(levelManager).Field<LevelSequence>("levelSequence").Value;
-            var originals = Traverse.Create(levelSequence).Field<string[]>("levels").Value;
-            var newlist = new List<string>
-            {
-                originals[0], // We need the Entrance to match the current one or the game will crash when moving to the next level.
-            };
-            newlist.AddRange(replacements); // Append user supplied levels
-            Traverse.Create(levelSequence).Field<string[]>("levels").Value = newlist.ToArray();
-            return originals.ToList().GetRange(1, 5);
+            var gsmLevelSequence = Traverse.Create(gameContext.gameStateMachine).Field<LevelSequence>("levelSequence").Value;
+            var originalSequence = Traverse.Create(gsmLevelSequence).Field<string[]>("levels").Value;
+            Traverse.Create(gsmLevelSequence).Field<string[]>("levels").Value =
+                replacements.Prepend(originalSequence[0]).ToArray();
+            return originalSequence.ToList();
```

Comment:
After some investigation, I found that the `levelSequence` from `levelManager` is different than the `levelSequence` from `gameContext`.

`levelManager.levelSequence` is set as part of the game creation procedure.

`gameContext.levelSequence` is more of a temporary field for the purposes of remembering which implied levelSequence the user selected to play.  This is the value that is then passed into `levelManager.levelSequence` during game creation.

By changing `OnPostGameCreated` to `OnPre...`, we could set `gameContext.levelSequence` directly.  This also allows us to get around the "first-level-must-match-game-type" limitation.  For example, I set RK entrance when playing EQ campaigns.

However, one limitation that still applies is that _some_ entrance must be the first level, even if it doesn't match the campaign.  As such, I've left the original "keep-the-first-entrance" logic, but did adjust this so as to apply during `OnPre...`.

---

Additionally, one issue I encountered was that the game would seem to crash at the end of the game.  While not 100% debugged, the issue seemed to be that we were replacing back the original values after deactivation (as we do with every other rule, of course).

It seems like the game didn't like that (perhaps because post-game the level sequence is supposed to be cleared?  I don't know - I didn't really dig).  Either way, _not replacing_ after deactivation fixed this.  Which means not only less code, but seems safe to do, as the levelSequence is discarded after end of game anyhow.  And changing level sequence _during_ matches once they've begun if the host leaves might have unintended consequences.

Reply:
Well done for cracking this one. The LevelSequence from GSM was returning valid level sequences for me, but changes were not getting applied. Thanks for investigating and fixing.